### PR TITLE
feat: LLVM pretty syntax based on macro elaboration

### PR DESCRIPTION
**Pull request by bollu**
This revives our old way of building custom syntax, by elaborating pretty syntax into generic syntax. Thanks @alexkeizer and @tobiasgrosser for the elbow grease :)

### CODE REVIEW

File: SSA/Projects/InstCombine/LLVM/PrettyEDSL.lean [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+def pretty_test :=
+  [alive_icom ()|{
+  ^bb0(%arg0: i32):
+    %0 = llvm.mlir.constant 8 : i32
+    %1 = llvm.add %0, %arg0 : i32
+    %2 = llvm.mul %1, %arg0 : i32
+    %3 = llvm.not %2 : i32
+    llvm.return %3 : i32
+  }]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/LLVM/PrettyEDSL.lean:64]: Maybe add a generic variant of this test too?

```suggestion

def pretty_test_generic (w : Nat) :=
  [alive_icom (w)|{
  ^bb0(%arg0: _):
    %0 = llvm.mlir.constant 8 : _
    %1 = llvm.add %0, %arg0 : _
    %2 = llvm.mul %1, %arg0 : _
    %3 = llvm.not %2 : _
    llvm.return %3 : _
  }]
  }]
```

File: SSA/Projects/InstCombine/LLVM/PrettyEDSL.lean [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+macro resName:mlir_op_operand " = " name:InstCombine.bin_op_name
+    x:mlir_op_operand ", " y:mlir_op_operand " : " t:mlir_type : mlir_op  => do
+  let name ← match name.raw with
+    | .node _ _ ⟨.atom _ name :: _⟩ => pure name
+    | _ => Macro.throwErrorAt name s!"Expected an atom, found: {name}" --TODO: better error message
+  let opName := Syntax.mkStrLit name
+  `(mlir_op| $resName:mlir_op_operand = $opName ($x, $y) : ($t, $t) -> ($t) )
+
+set_option hygiene false in
+macro res:mlir_op_operand " = " "llvm.mlir.constant" x:num " : " t:mlir_type : mlir_op =>
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/LLVM/PrettyEDSL.lean:52]: To double check, do we still need hygiene false in this case? Or can this be removed after your recent hygiene-related discoveries?

**Review by tobiasgrosser — APPROVED** [review]
Hi @alexkeizer, wow, this change makes our use interface really beautiful. I also checked the error messages and they are actually pretty decent. Very nice!

**Comment by bollu** [comment]
LGTM, @alexkeizer feel invited to merge at your leisure.

File: SSA/Projects/InstCombine/LLVM/PrettyEDSL.lean (hunk 2 above)
Comment (alexkeizer) [reply SSA/Projects/InstCombine/LLVM/PrettyEDSL.lean:52]: Indeed, this option can be removed now that our idents are no longer affected by hygiene

File: SSA/Core/MLIRSyntax/GenericParser.lean [hunk 3]
```diff
@@ -294,25 +294,25 @@ syntax "!" ident : mlir_type
 syntax ident: mlir_type
 syntax "_" : mlir_type
 
-
```
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/GenericParser.lean:297]: Unrelated change.

File: SSA/Core/MLIRSyntax/GenericParser.lean [hunk 4]
```diff
@@ -487,6 +487,7 @@ macro_rules
    let opsList <- `([mlir_ops| $ops])
    `(Region.mk "entry" [] $opsList)
 
+-- TODO: see if we can hide the `[mlir_op| ` in the same way as we did here for regions
```
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/GenericParser.lean:490]: Was this addressed in https://github.com/opencompl/ssa/pull/361?

Comment (alexkeizer) [reply SSA/Core/MLIRSyntax/GenericParser.lean]: It's actually already addressed here, I've dropped the TODO

**Comment by alexkeizer** [comment]
I've rebased on main (with #361 actually merged into it), made some final cleanups and hit merge
